pkg/api/validate/dynamic/dynamic.go

```diff
@@ -607,34 +612,80 @@ func (dv *dynamic) ValidateSubnets(ctx context.Context, oc *api.OpenShiftCluster
 			)
 		}
 
-		if oc.Properties.ProvisioningState == api.ProvisioningStateCreating {
-			if ss.SubnetPropertiesFormat != nil &&
-				ss.SubnetPropertiesFormat.NetworkSecurityGroup != nil {
-				expectedNsgID, err := subnet.NetworkSecurityGroupID(oc, s.ID)
+		subnetByID[s.ID] = ss
+	}
+	return subnetByID, nil
+}
+
+// checkByoNSG checks whether all the subnets have or don't have NSG attached.
+// when the BYONsg feature flag is on and only some of the subnets are attached with an NSG,
+// it returns an error.  If none of the subnets is attached, it's no longer BYO NSG and the
+// cluster installation process should fall back to using the managed nsg.
+func (dv *dynamic) checkByoNSG(subnetByID map[string]*mgmtnetwork.Subnet) (bool, error) {
+	noNSGAttached := 0
+	allSubnetsAreAttached := len(subnetByID)
+
+	var attached int
+	for _, subnet := range subnetByID {
+		if subnet.NetworkSecurityGroup != nil && subnet.NetworkSecurityGroup.ID != nil {
+			attached++
+		}
+	}
+	if attached > noNSGAttached && attached < allSubnetsAreAttached {
+		dv.log.Info("BYO NSG: not all subnets are attached")
+		return false,
+			&api.CloudError{
+				StatusCode: http.StatusBadRequest,
+				CloudErrorBody: &api.CloudErrorBody{
+					Code:    api.CloudErrorCodeInvalidLinkedVNet,
+					Message: errMsgNSGNotProperlyAttached,
+				},
+			}
+	}
+	if attached == allSubnetsAreAttached {
+		dv.log.Info("all subnets are attached, BYO NSG")
```

Comment:
This has  "fallback" logic in it.  I would prefer not to switch on this.  Let's talk with Puneet as if a customer specifies an API option or feature flag today but doesn't have everything in place we error out with a relevant error message.  

How will this work when we have an API option?  I would prefer the simpler approach of erroring out.

Reply:
IMO - Since we're toggling controllers internally it would be good to error on this so the customer knows why their feature isn't working.